Adds crash handling and reporting to GAPIC.

baldwinn860 opened the pull request:
This change adds an UncaughtExceptionHandler implementation that uploads a Throwable's
StackTrace to Crash2 for processing. Currently this targets the staging report area
which only accepts crashes from internal IP addresses and in effect is disabled for all
external users.

ben-clayton commented on gapic/src/main/com/google/gapid/util/Crash2ExceptionHandler.java:
```diff
@@ -0,0 +1,65 @@
+package com.google.gapid.util;
```
Missing copyright header

baldwinn860 replied:
Ah, that would be important, fixed.

ben-clayton commented:
Crash reports need to be enabled by the user via a setting.

pmuetschard commented on gapic/src/main/com/google/gapid/util/Crash2ExceptionHandler.java:
```diff
@@ -0,0 +1,80 @@
+ */
+public class Crash2ExceptionHandler implements Thread.UncaughtExceptionHandler {
+  private static final String CRASH_REPORT_PRODUCT = "GAPID";
+  private static final String CRASH_REPORT_VERSION = "Client:" + Version.GAPID_VERSION.toString();
```
You need to sync, this won't build once rebased on top of master. (you need to use GapidVersion.GAPID_VERSION, which is best statically import to not have to be so verbose).

pmuetschard commented on gapic/src/main/com/google/gapid/util/Crash2ExceptionHandler.java:
```diff
@@ -0,0 +1,80 @@
+  private static final String CRASH_REPORT_VERSION = "Client:" + Version.GAPID_VERSION.toString();
+  // TODO(baldwinn860): Send to production url when we get approval.
+  private static final String CRASH_REPORT_URL_BASE = "https://clients2.google.com/cr/staging_report?";
+  private static final String CRASH_REPORT_URL_FIELDS = "product=" + CRASH_REPORT_PRODUCT + "&version=" + CRASH_REPORT_VERSION;
```
The field values should be URL encoded, at least the version.

pmuetschard commented on gapic/src/main/com/google/gapid/Main.java:
```diff
@@ -47,6 +48,12 @@
+      Crash2ExceptionHandler handler = new Crash2ExceptionHandler(Thread.getDefaultUncaughtExceptionHandler());
+      Thread.setDefaultUncaughtExceptionHandler(handler);
+    }
+
```
Move this into a static function inside the C2EH class.

pmuetschard commented on gapic/src/main/com/google/gapid/util/Crash2ExceptionHandler.java:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  @Override
+  public void uncaughtException(Thread t, Throwable th) {
```
This isn't go. Please use descriptive variable names.

pmuetschard replied:
This should be done on a separate thread. Use Scheduler.EXECUTOR

pmuetschard commented on gapic/src/main/com/google/gapid/util/Crash2ExceptionHandler.java:
```diff
@@ -0,0 +1,80 @@
+  public void uncaughtException(Thread t, Throwable th) {
+    ByteArrayOutputStream traceStream = new ByteArrayOutputStream();
+    th.printStackTrace(new PrintStream(traceStream));
+    String traceString = traceStream.toString();
```
Use Throwables.getStackTraceAsString (you can then inline it, since it's only used in one place)

pmuetschard commented on gapic/src/main/com/google/gapid/util/Crash2ExceptionHandler.java:
```diff
@@ -0,0 +1,80 @@
+      // Creates a connection to crash2
+      OkHttpClient client = new OkHttpClient();
+      RequestBody body = new MultipartBuilder()
+        .type(MultipartBuilder.FORM)
```
nit: continuation indents should be four spaces.

pmuetschard commented on gapic/src/main/com/google/gapid/util/Crash2ExceptionHandler.java:
```diff
@@ -0,0 +1,80 @@
+  private static final String CRASH_REPORT_URL = CRASH_REPORT_URL_BASE + CRASH_REPORT_URL_FIELDS;
+  private static final Logger LOG = Logger.getLogger(Crash2ExceptionHandler.class.getName());
+
+  private Thread.UncaughtExceptionHandler oldDefaultHandler_;
```
No underscores or "m", "my", "our", whatever, prefixes/postfixes, please.

Also this can be final.

pmuetschard commented on gapic/src/main/com/google/gapid/util/Crash2ExceptionHandler.java:
```diff
@@ -0,0 +1,80 @@
+        .build();
+      Request request = new Request.Builder()
+        .url(CRASH_REPORT_URL)
+        .post(body)
```
Since you are not using the body anywhere else, you can inline the builder - c.f. proto builders.

pmuetschard commented on gapic/src/main/com/google/gapid/util/Crash2ExceptionHandler.java:
```diff
@@ -0,0 +1,80 @@
+        .build();
+      Response response = client.newCall(request).execute();
+      if (response.isSuccessful()) {
+        LOG.log(Level.INFO, "Crash Report Uploaded Successfully with " + response.code() +";\nCrash Report ID: " + response.body().string());
```
No newlines in log messages, please. It messes up the formatting.

baldwinn860 commented:
I believe I've addressed all the review comments @pmuetschard and @ben-clayton. I have the settings class disabling crash reporting until I can get the UI for enabling it together in another pull request.

pmuetschard commented on gapic/src/main/com/google/gapid/util/Crash2ExceptionHandler.java:
```diff
@@ -0,0 +1,106 @@
+
+  private final Thread.UncaughtExceptionHandler previousHandler;
+
+  public static void Register() {
```
lowercase method names

pmuetschard commented on gapic/src/main/com/google/gapid/util/Crash2ExceptionHandler.java:
```diff
@@ -0,0 +1,106 @@
+
+  @Override
+  public void uncaughtException(Thread thread, Throwable thrown) {
+    Scheduler.EXECUTOR.submit(new Runnable() {
```
Since you don't return a result and ignore the returned future, simply use `execute`.

pmuetschard replied:
Also, because this can happen at any time, ignore my previous suggestion to use Scheduler.EXECUTOR and just create a thread manually.

Do we need some rate control?

pmuetschard commented on gapic/src/main/com/google/gapid/util/Crash2ExceptionHandler.java:
```diff
@@ -0,0 +1,106 @@
+          LOG.log(Level.SEVERE, "Unable to upload exception to crash2", e);
+        } finally {
+          // Pass the exception back to the os to get logged
+          previousHandler.uncaughtException(thread, thrown);
```
Since this is now happening on a different thread, it's probably better to just invoke it after scheduling the crash2 stuff.

baldwinn860 commented:
One problem I realized while addressing the feedback is that any call to `Scheduler.EXECUTOR`s `submit` or `schedule` will never hit this `UncaughtExceptionHandler` as `ListenableFuture`s actually wrap their `Runnable`s in a catch block to transform it into a failure return value instead. One possible fix is to add a `onFailure` callback to all futures returned from `submit` and `schedule` calls, is this an acceptable fix @pmuetschard, or do you have another idea?

pmuetschard commented:
No, not every exception should be reported. Only the "unhandled" ones. This was my point when we were chatting. As is, this CL doesn't actually really do anything (even if we change the settings to enable reporting), because the RPC threads, background threads and the UI thread don't leak exceptions like this (or shouldn't). We have to hook the reporting into a few places where we currently just log or display the exception to the user, but have no better way of handling it.

Since I know best where those places are, please address all the review comments and make the reporting a utility function and I will hook it into the places where we currently just "ignore" unhandled exceptions.
